projects/01-rubygems-cli/solution/spec/lib/program_spec.rb

```diff
@@ -0,0 +1,133 @@
+require "spec_helper"
+require "./lib/program"
+require "./lib/command_results/command_error_result"
+require "./lib/command_results/show_command_result"
+require "./lib/command_results/search_command_result"
+
+RSpec.describe Program do
+  subject(:program) { described_class }
+
+  describe ".execute" do
+    subject(:execute) { program.execute(argv)}
+
+    context "with non existing command name" do
+      let(:argv) { ["non_existing_command"] }
+
+      it "returns CommandErrorResult" do
+        result = execute
+
+        expect(result.class).to be CommandErrorResult
+        expect(result.error_message).to eq "Invalid option non_existing_command!"
+        expect(result.exit_code).not_to eq 0
+      end
+    end
+
+    context "with nil command name" do
+      let(:argv) { [nil] }
+
+      it "returns CommandErrorResult" do
+        result = execute
+        
+        expect(result.class).to be CommandErrorResult
+        expect(result.error_message).to eq "No command provided!"
+        expect(result.exit_code).not_to eq 0
+      end
+    end
+
+    context "with empty command name" do
+      let(:argv) { [""] }
+
+      it "returns CommandErrorResult" do
+        result = execute
+        
+        expect(result.class).to be CommandErrorResult
+        expect(result.error_message).to eq "No command provided!"
+        expect(result.exit_code).not_to eq 0
+      end
+    end
+
+    context "with empty argv array" do
+      let(:argv) { [] }
+
+      it "returns CommandErrorResult" do
+        result = execute
+        
+        expect(result.class).to be CommandErrorResult
+        expect(result.error_message).to eq "No command provided!"
+        expect(result.exit_code).not_to eq 0
+      end
+    end
+
+    context "with show existing gem" do
+      let(:argv) { ["show", "rspec"] }
+
+      it "returns ShowCommandResult" do
+        result = execute
+
+        expect(result.class).to be ShowCommandResult
+        expect(result.gem_name).to eq "rspec"
+        expect(result.info).to eq "BDD for Ruby"
+        expect(result.exit_code).to eq 0
+      end
+    end
+
+    context "with show inexisting gem" do
+      let(:argv) { ["show", "non_existing_gem"] }
+
+      it "raises GemNotFound" do
+        # TODO: catch exception in ShowCommand
+        expect { execute }.to raise_error(GemNotFound)
```

Comment:
Still a todo because you are planning to return a CommandResult here as well ?

Reply:
96dbab46bb6f1986765b91ce04e57d82004b938a